### PR TITLE
ipvs: add the support of fnat46.

### PR DESCRIPTION
**Pull request by andrewhit**
- Introduce "xoa" for toa mode and uoa mode as below in dpvs.conf.
  tcp {
      toa_mode {normal|extra|xoa}
	  ...
  }

  uoa {
     uoa_mode {opp|xoa}  # "ipo" mode is removed.
  }

  Also, "extra" as one toa mode is introduced for ipv4, that includes
  (cip/cport, vip/vport) in tcp option.

- Consolidate the inbound and outbounding processing for
  fnat44/fnat66/fnat46/fnat64.

### CODE REVIEW

File: src/ipvs/ip_vs_xmit.c [hunk 1]
```diff
@@ -1520,10 +1460,10 @@ static int __dp_vs_out_xmit_snat4(struct dp_vs_proto *proto,
...
         fl4.fl4_saddr = conn->vaddr.in;
         fl4.fl4_tos = iph->type_of_service;
-        
+
 
         if (conn->outwall) {
-	    rt = route_gfw_net_lookup(&conn->caddr.in);
+        rt = route_gfw_net_lookup(&conn->caddr.in);
             if (!rt) {
                 err = EDPVS_NOROUTE;
```
Comment (ywc689) [inline src/ipvs/ip_vs_xmit.c:1468]: fix the indent issue

**Comment by ywc689** [comment]
The kernel module part for xoa is missing, and the uoa kernel module should be modified accordingly. I don't think it's a complete work now. Could you please replenish it?

File: include/ipvs/xoa.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+void *dp_vs_xoa_insert(struct rte_mbuf *mbuf, void *iph,
+                       int iptot_len, int iphdr_len, int xoa_len);
+void dp_vs_xoa4_fill(struct dp_vs_xoa_hdr *xoah, int af,
+                     union inet_addr *saddr, union inet_addr *daddr,
+                     uint16_t sport, uint16_t dport, bool is_asym_trans);
+void dp_vs_xoa6_fill(struct dp_vs_xoa_hdr *xoah, int af,
+                     union inet_addr *saddr, union inet_addr *daddr,
+                     uint16_t sport, uint16_t dport,
+                     uint8_t next_proto, bool is_asym_trans);
+int dp_vs_xoa_length(int af);
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply include/ipvs/xoa.h:138]: Removed.

File: src/ipv6/ipv6.c [hunk 3]
```diff
@@ -373,8 +373,7 @@ int ip6_output(struct rte_mbuf *mbuf)
     iftraf_pkt_out(AF_INET6, mbuf, dev);
     if (unlikely(conf_ipv6_disable)) {
         IP6_INC_STATS(outdiscards);
-        if (rt)
-            route6_put(rt);
+		route6_put(rt);
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/ipv6/ipv6.c:376]: Done.

File: src/ipv6/route6_lpm.c [hunk 4]
```diff
@@ -224,8 +224,7 @@ static struct route6 *rt6_lpm_input(const struct rte_mbuf *mbuf, struct flow6 *f
     return rt6;
 
 miss:
-    if (rt6)
-        route6_put(rt6);
+	route6_put(rt6);
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/ipv6/route6_lpm.c:227]: Done.

File: src/ipv6/route6_lpm.c [hunk 5]
```diff
@@ -245,8 +244,7 @@ static struct route6 *rt6_lpm_output(const struct rte_mbuf *mbuf, struct flow6 *
...
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/ipv6/route6_lpm.c:247]: Done.

File: src/ipvs/ip_vs_xoa.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        return true;
+    }
+
+    return false;
+}
+
+/*
+ * "xoa" is only applied to fnat46 and fnat66 modes.
+ */
+int dp_vs_xoa_get_iplen(struct dp_vs_conn *conn, struct rte_mbuf *mbuf,
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/ipvs/ip_vs_xoa.c:49]: XOA in both fnat44 and fnat64 modes uses the customized IP option that is unfriendly to the switch. And the switch normally drops this kind of packets in the case of high pps.

File: src/ipvs/ip_vs_proto_tcp.c [hunk 7]
```diff
@@ -690,6 +834,34 @@ tcp_conn_lookup(struct dp_vs_proto *proto, const struct dp_vs_iphdr *iph,
...
+        goto out;
+    }
+
+    /* add toa to syn packet */
+    if (th->syn && !th->ack) {
+        tcp_in_add_xoa(conn, mbuf);
+        goto out;
+    }
+
+    /* add toa to first data packet */
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/ipvs/ip_vs_proto_tcp.c:853]: Done.

File: src/ipvs/ip_vs_proto_udp.c [hunk 8]
```diff
@@ -416,78 +484,84 @@ static int send_standalone_uoa(const struct dp_vs_conn *conn,
...
-    if ((ip4_hdrlen(mbuf) + sizeof(struct ipopt_uoa) >
-                sizeof(struct iphdr) + MAX_IPOPTLEN)
-            || (mbuf->pkt_len + sizeof(struct ipopt_uoa) > mtu))
+    iphdrlen = iptot_len = xoa_len = 0;
+
+    err = dp_vs_xoa_get_iplen(conn, mbuf,
+                              &iphdrlen, &iptot_len, &xoa_len, mtu);
+    if (err != EDPVS_OK) {
         goto standalone_uoa;
+    }
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/ipvs/ip_vs_proto_udp.c:506]: - UDP in fnat44 uses the "opp" that is fine to the switch;
- UDP in fnat64  has two choices: a) xoa: that is unfriendly to the switch; 2) opp is not supported.

File: src/ipvs/ip_vs_proto_udp.c [hunk 9]
```diff
@@ -43,8 +45,10 @@ enum uoa_state {
 };
 
 enum uoa_mode {
-    UOA_M_OPP,      /* priave "option-protocol" (IPPROTO_OPT) with UOA */
-    UOA_M_IPO,      /* add UOA as IPv4 Option field */
+    UOA_M_OPP, /* ipv4 and ipv6: [cip/cport] included in private
+                  "option-protocol" (IPPROTO_OPT) */
+    UOA_M_XOA, /* ipv4 and ipv6: [cip/cport, vip/vport] included in ipv4 option
+                  or ipv6 dst option */
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/ipvs/ip_vs_proto_udp.c:51]: As for the kernel module, toa/uoa, please refer to the below links.

https://github.com/bytedance/toa
https://github.com/bytedance/uoa

File: src/route.c [hunk 10]
```diff
@@ -618,10 +617,10 @@ static int route_sockopt_get(sockoptid_t opt, const void *conf, size_t size,
...
 
     nroute = rte_atomic32_read(&this_num_routes);
-    
+
     if (cf && cf->outwalltb) {
-	nroute = rte_atomic32_read(&this_num_out_routes);
-	outwall_table = true;
+    nroute = rte_atomic32_read(&this_num_out_routes);
+    outwall_table = true;
     }
```
[the remark this answers is not in the document]
Comment (andrewhit) [reply src/route.c]: Done.

**Comment by andrewhit** [comment]
Closing this PR because dpvs has supported this feature.

**Pull request closed without merging**